### PR TITLE
Support Selenium 3.0.1

### PR DESCRIPTION
**Pull request by krmahadevan**
Adding changes for 3.0.1. Also bumped up the Java
version to JDK8 because Selenium 3 now requires JDK8.

### CODE REVIEW

**Comment by foxhound91** [comment]
The Jenkins java version is 1.7 and we need to build plugins for Java 8
> JAVA_HOME	/opt/jdk/jdk1.7.latest

@Mobrockers Could you please ammend that?

**Comment by Mobrockers** [comment]
@foxhound91 I'll look into it. Things seem to have changed a bit since I last worked on this so I am unsure where to request this. It is not something I can change myself.

**Comment by Mobrockers** [comment]
Could you please add to the readme that this plugin requires Java 8? Java 8 builds should be enabled so on adding this commit the build should also be retried.

File: pom.xml [hunk 1]
```diff
@@ -16,9 +16,9 @@
                 this result is useless to us -->
         <findbugs.failOnError>false</findbugs.failOnError>
 
-		<selenium.short.version>2.53</selenium.short.version>
-		<selenium.version>${selenium.short.version}.0</selenium.version>
-        <htmlunit.version>2.20</htmlunit.version>
+		<selenium.version>3.0.1</selenium.version>
```
Comment (Mobrockers) [inline pom.xml]: Can you please fix all your indents in this file

Comment (krmahadevan) [inline pom.xml]: @Mobrockers - I have fixed it.

**Comment by Mobrockers** [comment]
There are test failures, could you please fix these?

**Comment by krmahadevan** [comment]
@Mobrockers - Sure thing. I don't have a lot of experience working with Jenkins plugins. So I am not able to figure out where to start with the tests that are failing. The tests are failing with `CookieConfig cannot be set after ServletContext is started`. But my changes are not involved in anything around that. To make it more worse, since I don't know what Jenkins uses under the hoods to start off a server when one does `java -jar jenkins.war` its becoming difficult for me to understand what could be going wrong. It would be great if you could please help me with this issue so that I can work on wrapping up this PR.

**Comment by Mobrockers** [comment]
Those are just warnings, you can ignore those. The actual errors are these two at the end of the test run:

Results :

Tests in error: 
org.junit.runner.manipulation.Filter.initializationError(org.junit.runner.manipulation.Filter)
  Run 1: Filter.initializationError »  No tests found matching Method org.jvnet.hudson....
  Run 2: Filter.initializationError »  No runnable methods
  Run 3: Filter.initializationError »  No runnable methods
  Run 4: Filter.initializationError »  No runnable methods

  FailedTest.org.jvnet.hudson.test.JellyTestSuiteBuilder$JellyTestSuite » NoSuchMethod


Which seem to be coming from the jenkins test harness (https://github.com/jenkinsci/jenkins-test-harness), but I do not have enough experience with jenkins plugins either to determine where to start looking. I also do not have the time at the moment to investigate. Perhaps it is as simple as updating the jenkins test framework versions in the pom?

**Comment by krmahadevan** [comment]
@Mobrockers - I finally managed to fix the build failures. I guess we can now go ahead and get this merged.

**Comment by Mobrockers** [comment]
I appreciate your effort but this does not seem like an acceptable fix to the issue. Simply adding a missing dependency as a class to the project is a no-go. I really do appreciate the effort though. If I can find some spare time I can take a look at a proper fix, or you can try to fix it properly yourself, but it's not getting merged with that deprecated class simply added back in as code. 

Since the class is deprecated, there must be a replacement no? Perhaps all that's needed is to replace that dependency with its replacement.

**Comment by krmahadevan** [comment]
I don't think that's the case. The class that was added is referring to a method called closeAllWindows() which seems to exist in an overloaded version of WebClient that exists within Jenkins codebase. When I bump selenium version to 3.1.0 that is getting shadowed out by the original class via selenium. All I did was to duplicate the class and remove reference to the non existent method. The deprecated class is being called from test harness. So not much I can do to change things there. I leave it to you to decide what you want to do with this pull request.

**Comment by Mobrockers** [comment]
Okay, apologies. Interesting find, if you had included it as a comment or perhaps on the commit I would've understood better why the class was added this way.

Do you think it might be possible to explicitly load the WebClient dependency last in the pom, so that the selenium version is not used for the test phase? Have you tried this already?

**Comment by krmahadevan** [comment]
I did include it in my commit message "Adding changes for 3.1.0. Also bumped up the Java
version to JDK8 because Selenium 3 now requires JDK8.

Had to resort to duplicating "org.jvnet.hudson.test.HudsonTestCase"
because this class refers to a method which is no longer existing."

I tried various different things. Since selenium dependency is a compile time dependency and since test harness comes in via the parent pom I didn't make much progress. Eventually I had to settle for this approach of duplicating the class.

**Comment by Mobrockers** [comment]
'Had to resort to duplicating "org.jvnet.hudson.test.HudsonTestCase"
because this class refers to a method which is no longer existing."'

This does not explain why that is an issue though, it does not explain that it is a dependency conflict between selenium and the test harness. Which is very useful information to me (and anyone else looking at this weird class being in this project in the future), to understand your reason for settling for the chosen approach.

In any case, I will think on it and ask for feedback from jenkins devs on how to proceed. I'll see if I can make time this weekend, since you put in so much effort.

**Comment by krmahadevan** [comment]
👍

**Comment by krmahadevan** [comment]
@Mobrockers - I revisited this. This time, I was able to get rid of the duplicated class and still get the tests to pass by merely changing the order of dependencies between `htmlunit-driver-standalone` and `selenium-server-standalone` in the pom file. I guess now this should be good to be merged.

**Pull request merged**